### PR TITLE
DEP: remove allocation_tracking, deprecate PyDataMem_SetEventHook 

### PR DESCRIPTION
**Pull request by mattip**
Alternative to #19399. Rather than improving the documentation, lets remove this tool.

### CODE REVIEW

File: doc/release/upcoming_changes/20394.deprecation.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Removed outdated tool/allocation_tracking
+-----------------------------------------
+
+The ability to track allocations is now built-in to python.
```
Comment (seberg) [inline doc/release/upcoming_changes/20394.deprecation.rst:4]:
```suggestion
The ability to track allocations is now built-in to python using tracemalloc.
```
Just thought to add the term to use in search engines :).

File: numpy/core/src/multiarray/alloc.c [hunk 2]
```diff
@@ -218,6 +238,11 @@ PyDataMem_SetEventHook(PyDataMem_EventHookFunc *newhook,
     PyDataMem_EventHookFunc *temp;
     NPY_ALLOW_C_API_DEF
     NPY_ALLOW_C_API
+    /* 2021-11-18, 1.23 */
+    WARN_NO_RETURN(PyExc_DeprecationWarning,
+                     "PyDataMem_SetEventHook is deprecated, use tracemalloc "
+                     "and the 'np.lib.tracemalloc_domain' domain");
+    PyErr_Clear();
```
Comment (seberg) [inline numpy/core/src/multiarray/alloc.c:245]:
```suggestion
```
No need, not that it matters.

File: numpy/core/src/multiarray/alloc.c [hunk 3]
```diff
@@ -435,31 +460,13 @@ PyDataMem_UserNEW_ZEROED(size_t nmemb, size_t size, PyObject *mem_handler)
...
 
 NPY_NO_EXPORT void
 PyDataMem_UserFREE(void *ptr, size_t size, PyObject *mem_handler)
 {
     PyDataMem_Handler *handler = (PyDataMem_Handler *) PyCapsule_GetPointer(mem_handler, "mem_handler");
     if (handler == NULL) {
-        WARN_IN_FREE(PyExc_RuntimeWarning,
+        WARN_NO_RETURN(PyExc_RuntimeWarning,
                      "Could not get pointer to 'mem_handler' from PyCapsule");
         PyErr_Clear();
```
Comment (seberg) [inline numpy/core/src/multiarray/alloc.c:471]:
```suggestion
```
Seems I didn't notice it here :).

**Review by seberg — APPROVED** [review]
Thanks looks good to me, all comments are pretty much nits that don't matter much.  I am planning to merge it (maybe with the suggestions applied) tomorrow, if nobody beats me to it.

It seems right that tracemalloc is strictly better and this should affect few users.

File: numpy/core/src/multiarray/alloc.c (hunk 3 above)
Comment (mattip) [reply numpy/core/src/multiarray/alloc.c]: done

File: doc/release/upcoming_changes/20394.deprecation.rst (hunk 1 above)
Comment (mattip) [reply doc/release/upcoming_changes/20394.deprecation.rst]: done

**Comment by mattip** [comment]
Mailing list thread for the deprecation https://mail.python.org/archives/list/numpy-discussion@python.org/thread/2ADDZE7GMTUYHPUJ5TR4FGIALMVBBL3K/

**Comment by seberg** [comment]
Thanks, Matti, nice release note upgrade!  Going to be the optimist with respect to the mailing list thread, expect silence, and just put it in :).

**Pull request merged**